// @paper-design/shaders-react/dist/shaders/grain-gradient.js
/* * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                    Paper Shaders                    *
 *       https://github.com/paper-design/shaders       *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * */

import { memo } from "react";
import { ShaderMount } from "../shader-mount";
import {
  getShaderColorFromString,
  grainGradientFragmentShader,
  ShaderFitOptions,
  defaultPatternSizing,
  defaultObjectSizing,
  GrainGradientShapes
} from "@paper-design/shaders";
import { jsx } from "react/jsx-runtime";
const defaultPreset = {
  name: "Default",
  params: {
    ...defaultPatternSizing,
    speed: 1,
    frame: 0,
    colors: ["#000a0f", "#c4730b", "#bdad5f", "#d8ccc7"],
    softness: 0.7,
    intensity: 0.15,
    noise: 0.5,
    shape: "wave"
  }
};
const dotsPreset = {
  name: "Dots",
  params: {
    ...defaultPatternSizing,
    scale: 0.6,
    speed: 1,
    frame: 0,
    colors: ["#0a0000", "#6f0000", "#0080ff", "#f2ebc9"],
    softness: 0.75,
    intensity: 0.15,
    noise: 0.7,
    shape: "dots"
  }
};
const truchetPreset = {
  name: "Truchet",
  params: {
    ...defaultPatternSizing,
    speed: 1,
    frame: 0,
    colors: ["#0a0000", "#6f2200", "#eabb7c", "#39b523"],
    softness: 0,
    intensity: 0.2,
    noise: 1,
    shape: "truchet"
  }
};
const cornersPreset = {
  name: "Corners",
  params: {
    ...defaultObjectSizing,
    speed: 1,
    frame: 0,
    colors: ["#031018", "#00aeff", "#00ffcc", "#ffc800"],
    softness: 0.4,
    intensity: 0.35,
    noise: 0.35,
    shape: "corners"
  }
};
const ripplePreset = {
  name: "Ripple",
  params: {
    ...defaultObjectSizing,
    scale: 0.5,
    speed: 1,
    frame: 0,
    colors: ["#140a00", "#6f2d00", "#88ddae", "#2c0b1d"],
    softness: 0.5,
    intensity: 0.5,
    noise: 0.5,
    shape: "ripple"
  }
};
const blobPreset = {
  name: "Blob",
  params: {
    ...defaultObjectSizing,
    scale: 1.3,
    speed: 1,
    frame: 0,
    colors: ["#0f0e18", "#3e6172", "#a49b74", "#568c50"],
    softness: 0,
    intensity: 0.15,
    noise: 0.5,
    shape: "blob"
  }
};
const spherePreset = {
  name: "Sphere",
  params: {
    ...defaultObjectSizing,
    speed: 1,
    frame: 0,
    colors: ["#000319", "#0059b3", "#37f5f5", "#18c039"],
    softness: 1,
    intensity: 0.15,
    noise: 0.5,
    shape: "sphere"
  }
};
const moonPreset = {
  name: "Moon",
  params: {
    ...defaultObjectSizing,
    scale: 0.6,
    speed: 1,
    frame: 0,
    colors: ["#000000", "#000000", "#28272d", "#ffeccc"],
    softness: 1,
    intensity: 0.56,
    noise: 1,
    shape: "sphere"
  }
};
const grainGradientPresets = [
  cornersPreset,
  defaultPreset,
  dotsPreset,
  truchetPreset,
  ripplePreset,
  blobPreset,
  spherePreset,
  moonPreset
];
const GrainGradient = memo(function GrainGradientImpl({
  // Own props
  speed = defaultPreset.params.speed,
  frame = defaultPreset.params.frame,
  colors = defaultPreset.params.colors,
  softness = defaultPreset.params.softness,
  intensity = defaultPreset.params.intensity,
  noise = defaultPreset.params.noise,
  shape = defaultPreset.params.shape,
  // Sizing props
  fit = defaultPreset.params.fit,
  scale = defaultPreset.params.scale,
  rotation = defaultPreset.params.rotation,
  originX = defaultPreset.params.originX,
  originY = defaultPreset.params.originY,
  offsetX = defaultPreset.params.offsetX,
  offsetY = defaultPreset.params.offsetY,
  worldWidth = defaultPreset.params.worldWidth,
  worldHeight = defaultPreset.params.worldHeight,
  ...props
}) {
  const uniforms = {
    // Own uniforms
    u_colors: colors.map(getShaderColorFromString),
    u_colorsCount: colors.length,
    u_softness: softness,
    u_intensity: intensity,
    u_noise: noise,
    u_shape: GrainGradientShapes[shape],
    // Sizing uniforms
    u_fit: ShaderFitOptions[fit],
    u_scale: scale,
    u_rotation: rotation,
    u_offsetX: offsetX,
    u_offsetY: offsetY,
    u_originX: originX,
    u_originY: originY,
    u_worldWidth: worldWidth,
    u_worldHeight: worldHeight
  };
  return /* @__PURE__ */ jsx(
    ShaderMount,
    {
      ...props,
      speed,
      frame,
      fragmentShader: grainGradientFragmentShader,
      uniforms
    }
  );
});
export {
  GrainGradient,
  blobPreset,
  cornersPreset,
  defaultPreset,
  dotsPreset,
  grainGradientPresets,
  moonPreset,
  ripplePreset,
  spherePreset,
  truchetPreset
};
