// @paper-design/shaders-react/dist/shaders/smoke-ring.js
/* * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                    Paper Shaders                    *
 *       https://github.com/paper-design/shaders       *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * */

import { memo } from "react";
import { ShaderMount } from "../shader-mount";
import { colorPropsAreEqual } from "../color-props-are-equal";
import {
  defaultObjectSizing,
  getShaderColorFromString,
  getShaderNoiseTexture,
  smokeRingFragmentShader,
  ShaderFitOptions
} from "@paper-design/shaders";
import { jsx } from "react/jsx-runtime";
const defaultPreset = {
  name: "Default",
  params: {
    ...defaultObjectSizing,
    speed: 0.4,
    frame: 0,
    colorBack: "#ffffff",
    colors: ["#136c5e", "#0f0224"],
    noiseScale: 5,
    noiseIterations: 10,
    radius: 0.5,
    thickness: 0.25,
    innerShape: 1.2
  }
};
const poisonPreset = {
  name: "Poison",
  params: {
    ...defaultObjectSizing,
    speed: 1,
    frame: 0,
    colorBack: "#000000",
    colors: ["#d4ff00", "#0c7f08"],
    noiseScale: 2.2,
    noiseIterations: 10,
    radius: 0.4,
    thickness: 0.2,
    innerShape: 0.6
  }
};
const linePreset = {
  name: "Line",
  params: {
    ...defaultObjectSizing,
    frame: 0,
    colorBack: "#000000",
    colors: ["#1fe8ff", "#4540a4"],
    noiseScale: 1.1,
    noiseIterations: 2,
    radius: 0.38,
    thickness: 0.01,
    innerShape: 0.88,
    speed: 4
  }
};
const cloudPreset = {
  name: "Cloud",
  params: {
    ...defaultObjectSizing,
    frame: 0,
    colorBack: "#3b9bff",
    colors: ["#ffffff"],
    noiseScale: 1.5,
    noiseIterations: 10,
    radius: 0.5,
    thickness: 0.65,
    innerShape: 0.9,
    speed: 0.5
  }
};
const smokeRingPresets = [defaultPreset, linePreset, poisonPreset, cloudPreset];
const SmokeRing = memo(function SmokeRingImpl({
  // Own props
  speed = defaultPreset.params.speed,
  frame = defaultPreset.params.frame,
  colorBack = defaultPreset.params.colorBack,
  colors = defaultPreset.params.colors,
  noiseScale = defaultPreset.params.noiseScale,
  thickness = defaultPreset.params.thickness,
  radius = defaultPreset.params.radius,
  innerShape = defaultPreset.params.innerShape,
  noiseIterations = defaultPreset.params.noiseIterations,
  // Sizing props
  fit = defaultPreset.params.fit,
  scale = defaultPreset.params.scale,
  rotation = defaultPreset.params.rotation,
  originX = defaultPreset.params.originX,
  originY = defaultPreset.params.originY,
  offsetX = defaultPreset.params.offsetX,
  offsetY = defaultPreset.params.offsetY,
  worldWidth = defaultPreset.params.worldWidth,
  worldHeight = defaultPreset.params.worldHeight,
  ...props
}) {
  const noiseTexture = typeof window !== "undefined" && { u_noiseTexture: getShaderNoiseTexture() };
  const uniforms = {
    // Own uniforms
    u_colorBack: getShaderColorFromString(colorBack),
    u_colors: colors.map(getShaderColorFromString),
    u_colorsCount: colors.length,
    u_noiseScale: noiseScale,
    u_thickness: thickness,
    u_radius: radius,
    u_innerShape: innerShape,
    u_noiseIterations: noiseIterations,
    ...noiseTexture,
    // Sizing uniforms
    u_fit: ShaderFitOptions[fit],
    u_scale: scale,
    u_rotation: rotation,
    u_offsetX: offsetX,
    u_offsetY: offsetY,
    u_originX: originX,
    u_originY: originY,
    u_worldWidth: worldWidth,
    u_worldHeight: worldHeight
  };
  return /* @__PURE__ */ jsx(ShaderMount, { ...props, speed, frame, fragmentShader: smokeRingFragmentShader, uniforms });
}, colorPropsAreEqual);
export {
  SmokeRing,
  cloudPreset,
  defaultPreset,
  linePreset,
  poisonPreset,
  smokeRingPresets
};
